crates/runtime/src/datafusion/cluster/codec/spice_physical_codec.rs

```diff
@@ -0,0 +1,139 @@
+        write!(f, "SpicePhysicalCodec")
+    }
+}
+
+impl SpicePhysicalCodec {
+    pub fn new_codec(runtime: Arc<Runtime>) -> Result<Arc<dyn PhysicalExtensionCodec>> {
+        Ok(Arc::new(Self {
+            inner: Arc::new(BallistaPhysicalExtensionCodec::default()),
+            runtime: Some(runtime),
+        }))
+    }
+
+    /// Used during encode and decode
+    fn runtime(&self) -> Result<Arc<Runtime>> {
+        self.runtime.clone().ok_or(DataFusionError::Execution(
+            "SpicePhysicalCodec did not bind a Runtime handle. This is a bug.".to_string(),
+        ))
+    }
+}
+
+impl PhysicalExtensionCodec for SpicePhysicalCodec {
+    fn try_decode(
+        &self,
+        buf: &[u8],
+        inputs: &[Arc<dyn ExecutionPlan>],
+        registry: &dyn FunctionRegistry,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        if let Ok(plan) = self.inner.try_decode(buf, inputs, registry) {
+            return Ok(plan);
+        }
+
+        let exec_params = serde_json::from_slice::<HashMap<String, Value>>(buf)
+            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+
+        match exec_params.get("spice.exec.name").and_then(|v| v.as_str()) {
+            Some("SchemaCastScanExec") => {
+                let schema = deserialize!(exec_params, "spice.exec.schema", Schema)?;
+                let exec = Arc::new(SchemaCastScanExec::new(
+                    Arc::clone(&inputs[0]),
+                    Arc::new(schema),
+                ));
+
+                Ok(exec)
+            }
+            Some("BytesProcessedExec") => {
+                // TODO: Make RequestContext serializable
+                Ok(Arc::new(BytesProcessedExec::new(Arc::clone(&inputs[0]))))
+            }
+            _ => exec_err!("Unsupported spice.exec.name"),
+        }
+    }
+
+    fn try_encode(&self, node: Arc<dyn ExecutionPlan>, buf: &mut Vec<u8>) -> Result<()> {
+        let mut map: HashMap<&str, Value> = HashMap::new();
+        map.insert("spice.exec.name", node.name().into());
+
+        match node.name() {
+            "SchemaCastScanExec" => {
+                let Some(concrete) = node.as_any().downcast_ref::<SchemaCastScanExec>() else {
+                    return exec_err!("Unable to serialize plan node");
+                };
+
+                map.insert(
+                    "spice.exec.schema",
+                    serde_json::to_value(concrete.schema())
+                        .map_err(|e| DataFusionError::Execution(e.to_string()))?,
+                );
+            }
+            "BytesProcessedExec" => { /* no-op */ }
+            "DataSourceExec" => {
+                let Some(concrete) = node.as_any().downcast_ref::<DataSourceExec>() else {
+                    return exec_err!("Unable to serialize plan node");
+                };
+
+                let data_source = concrete.data_source();
+
+                // Clearer error message instead of "unsupported plan"
+                if data_source
```

Comment:
`MemorySourceConfig` is literally a container for vecs of `RecordBatch`. We _could_ ship them to be crunched and then returned, but that seems like an antipattern.

You will see this when trying to run distributed queries with acceleration enabled.